### ISSUE TITLE
bin files left unprocessed by Distributed table

### ISSUE THREAD

**Issue by kmatt** [yandex/ClickHouse#1758]
I have a 4 node cluster running v1.1.54327-stable on CentOS 7.2.1511 using Distributed and ReplicatedMergeTree tables, with a simple insert loading data into the Distributed table (via https://github.com/mymarilyn/clickhouse-driver):

    CREATE TABLE test_shard ( ...)
    ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/cds', '{replica}', date, (pk), 8192);
    
    CREATE TABLE test (shard_key Int64, ...)
    ENGINE = Distributed(test_cluster, testdb, test_shard, shard_key);

Everything appears to be working as expected except there is a significant amount of disk space consumed by the Distributed table on the node accepting the inserts: 70GB for the Distributed table when each shard table uses 55GB disk. This remains many hours after a batch load completed, with no new inserts during that period.

.bin files appear to be unprocessed and consuming disk space in the Distributed table path, but no errors in the clickhouse-server.log.

Initial counts seem comparing table rows and source file lines seem to indicate that all inserts are succeeding, but the count of .bin files in the Distributed table directory grow with each file inserted, and those .bin files do not appear in the logs in /var/log/clickhouse-server #

**Comment by kmatt** [comment]
It looks like some of the .bin files left in directory for the Distributed table are not getting touched at all.

Picking some at random with a file time of 0000 GMT today (19 hours ago), 

    -rw-r----- 1 root root 205332 Jan 12 00:00 6350.bin

they do not appear as processed in the log:

    grep '/6350.bin`' clickhouse-server.log

... returns nothing for the current date.

When that same filename was part of a previous load yesterday, I see it being processed:

    2018.01.11 19:27:43.243860 [ 107 ] <Trace> cds.Distributed.DirectoryMonitor: Started processing `/localdata2/ch/data/logstore/cds/default@209%2E197%2E7%2E135:9000/6350.bin`
    2018.01.11 19:27:43.516764 [ 107 ] <Trace> cds.Distributed.DirectoryMonitor: Finished processing `/localdata2/ch/data/logstore/cds/default@209%2E197%2E7%2E135:9000/6350.bin`

**Comment by kmatt** [comment]
Its possible based on counts in table and from source files that all inserts are succeeding, but backlog of .bin files grows with new inserts, and those .bin files are not logged as successful or with error.

**Comment by ztlpn** [comment]
I couldn't reproduce it. Can you try to produce a minimal example that I can test? Also, please share your cluster config.

Some pointers: if you didn't restart ClickHouse, numbers in the names of temporary .bin files should be unique. Thus it is not possible that two data loads produce files with the same name. Maybe they are duplicated somehow? Another strange thing: `6350.bin` is owned by root. Do you run ClickHouse under the root user? That would be strange if ClickHouse couldn't delete the files that it itself created, but maybe there is a problem with file permissions.

**Comment by kmatt** [comment]
> Some pointers: if you didn't restart ClickHouse, numbers in the names of temporary .bin files should be unique.

Yes, Clickhouse engine was restarted between loads, so the sequence number would have restarted.

>  Do you run ClickHouse under the root user?

Yes, using this init.d script: https://gist.github.com/kmatt/5d41993f470e4d35315874285075e97f

    $ ps aux | grep clickhouse
    root     10703 21.4  0.6 1769816 845628 ?      Ssl  Jan11 1311:58 clickhouse-server --daemon --pid-file=/var/run/clickhouse-server/clickhouse-server.pid --config-file=/etc/clickhouse-server/config.xml

**Comment by kmatt** [comment]
Since I'm running under CentOS, and not using the Altinity RPMs due to issues with OpenSSL version compatibility, is there anything that could be wrong with the way Clickhouse is started on these nodes using the init.d script above?

**Comment by kmatt** [comment]
@ztlpn The server is only logging files processed to other nodes, the node I'm running ETL on is not mentioned:

```
[/var/log/clickhouse-server:#] awk -F'/' '/DirectoryMonitor/ { s[$7]++ } END { for (i in s) print i, s[i] }' clickhouse-server.log
default@xxx%2Exxx%2E7%2E123:9000 140764
default@xxx%2Exxx%2E7%2E122:9000 140770
default@xxx%2Exxx%2E7%2E135:9000 140772
```

.136 is the 4th node receiving the insert statements. Anything of interest in this?

**Comment by kmatt** [comment]
Of 15,586,402,301 lines in the source files, 15,584,535,754 appear in a COUNT(*) in Clickhouse. 

Close enough that I know that with 155GB per node on a 4 server cluster, the 200GB of remaining .bin files in the path for the Distributed table were largely ingested.

They just are not removed, and do not appear in the server log with the files that are removed.

**Comment by ztlpn** [comment]
Ok, so apparently this is the duplicate of #1404 (that was fixed in a4636f86563482ae37a6d92dfa5ff1361458b4b3 and released in 1.1.54337).

The bug is triggered when there is a shard with the single replica which is local and `internal_replication` is true (just as in your case). Do `.bin` files reside directly in the `/localdata2/ch/data/logstore/cds/` directory (not in its descendants)? If then, you can safely delete them^ they were already inserted.

To fix the bug you can upgrade to the newest version, but the simplest fix would be to just remove the `<internal_replication>` tag in the cluster config (as there is only one replica per shard, it doesn't do anything except trigger the bug).
